yeat/config/config.py

```diff
@@ -0,0 +1,70 @@
+# -------------------------------------------------------------------------------------------------
+# Copyright (c) 2022, DHS. This file is part of YEAT: http://github.com/bioforensics/yeat
+#
+# This software was prepared for the Department of Homeland Security (DHS) by the Battelle National
+# Biodefense Institute, LLC (BNBI) as part of contract HSHQDC-15-C-00064 to manage and operate the
+# National Biodefense Analysis and Countermeasures Center (NBACC), a Federally Funded Research and
+# Development Center.
+# -------------------------------------------------------------------------------------------------
+
+from . import PACBIO_READS, OXFORD_READS, READ_TYPES, AssemblyConfigError
+from .assembly import Assembly
+from .sample import Sample
+
+
+BASE_KEYS = ("samples", "assemblies")
+ASSEMBLY_KEYS = ("algorithm", "extra_args", "samples", "mode")
+
+
+class AssemblyConfig:
+    def __init__(self, data, threads):
+        self.data = data
+        self.validate()
+        self.threads = threads
+        self.create_sample_and_assembly_objects()
+        self.validate_samples_to_assembly_modes()
+
+    def validate(self):
+        self.check_required_keys(self.data.keys(), BASE_KEYS)
+        for key, value in self.data["samples"].items():
+            self.check_valid_keys(value.keys(), READ_TYPES)
+        for key, value in self.data["assemblies"].items():
+            self.check_required_keys(value.keys(), ASSEMBLY_KEYS)
+
+    def check_required_keys(self, observed_keys, expected_keys):
+        missing_keys = set(expected_keys) - set(observed_keys)
+        if missing_keys:
+            key_str = ",".join(sorted(missing_keys))
+            message = f"Missing assembly configuration setting(s) '{key_str}'"
+            raise AssemblyConfigError(message)
+        self.check_valid_keys(observed_keys, expected_keys)
+
+    def check_valid_keys(self, observed_keys, expected_keys):
+        extra_keys = set(observed_keys) - set(expected_keys)
+        if extra_keys:
+            key_str = ",".join(sorted(extra_keys))
+            message = f"Found unsupported configuration key(s) '{key_str}'"
+            raise AssemblyConfigError(message)
+
+    def create_sample_and_assembly_objects(self):
+        self.samples = {}
+        self.assemblies = {}
+        for key, value in self.data["samples"].items():
+            self.samples[key] = Sample(key, value)
+        for key, value in self.data["assemblies"].items():
+            self.assemblies[key] = Assembly(key, value, self.threads)
```

Comment:
Two comments.

- It's good practice to instantiate instance data in the constructor, even if you fill it in with a dedicated function like this. The code here technically works, but it's common to try to get a sense for an object's contents by looking at the constructor. And in this case, they'd miss `.samples` and `.assemblies`.
- Instead of "key" and "value", I'd use more descriptive variable names like "sample_name", "label", and "config".